### ISSUE TITLE
Add thread support

### ISSUE THREAD

**Issue by patcon** [42wim/matterbridge#638]
Just adding a place to track details for threading support across protocols.

- [x] Slack
- [ ] Zulip
- [ ] Mattermost

## Zulip

Should be possible to support threads as "topics".
https://zulipchat.com/api/send-message

**Comment by nylen** [comment]
For Discord (and other platforms that support message editing), it seems like it would be possible to edit thread replies into the original message.

**Comment by 42wim** [comment]
Discord has no threads. normal message editing already works

**Comment by nylen** [comment]
Yes, I know. I'm suggesting that for Discord, it would be possible to simulate threads by re-using the original message.

Example:

> This is a message in Discord

After a reply in a Slack thread, the same single message would look like this:

> This is a message in Discord
>
> [Thread] This is a thread reply

This would preserve the context of the thread in what seems to me like a natural way.

**Comment by Helcaraxan** [comment]
Although I understand where you are coming from with the threading not being carried over I don't think that the proposed solution achieves what you really want for both UX and technical reasons:
- Technically there are limits to message lengths in Discord. Which means long running threads will become problematic.
- From a pure UX perspective imagine that you are in Discord and observing a thread being created from Slack.
   - Now if you reply to it your response will be in Discord below the thread-edited message.
   - In Slack it will also be coming in as a fresh message as there is no way of knowing it was a reply to the thread.
   - If a thread participant on Slack notices your messages:
      - They can reply in the thread on Slack which in Discord will appear _above_ your message, which is really strange and unpleasant UX.
      - They can reply with a normal Slack message or start a new Slack thread on your message, which means your mechanic is destroying the logic behind Slack threads rendering them completely useless.

The later scenario is already the case nowadays because of the lack of an ability to recognise a Discord message as being targeted at a particular thread but it is not solved by this solution either.

As a side-note: I don't think that something like Matterbridge works well in bridging between threaded and non-threaded platforms, unless you mimic threading near-perfectly in the non-threaded one you will encounter issues.

**Comment by nylen** [comment]
>I don't think that the proposed solution achieves what you really want for both UX and technical reasons ... I don't think that something like Matterbridge works well in bridging between threaded and non-threaded platforms

Good points, I agree with you now.

**Comment by nylen** [comment]
Another idea for Discord: prefix `[thread]` to the message text when copying a threaded message from Slack.  This at least makes it clearer that there is additional context to the message.

I have working code for this if there's interest.

**Comment by patcon** [comment]
Heh #557 :)

**Comment by nylen** [comment]
Here's what I came up with, for Discord only:

```
diff --git a/bridge/discord/discord.go b/bridge/discord/discord.go
index a9c508c..08e45ac 100644
--- a/bridge/discord/discord.go
+++ b/bridge/discord/discord.go
@@ -152,6 +152,13 @@ func (b *Bdiscord) Send(msg config.Message) (string, error) {
 		msg.Text = "_" + msg.Text + "_"
 	}
 
+	if msg.ParentID != "" {
+		// This message is part of a thread
+		// Note: For this to work, `PreserveThreading` must be set to True on
+		// the [discord.servername] config block
+		msg.Text = "[thread] " + msg.Text
+	}
+
 	// use initial webhook configured for the entire Discord account
 	isGlobalWebhook := true
 	wID := b.webhookID
```

**Comment by ForsakenHarmony** [comment]
Is there currently any way to know if a message is coming from a thread? Just posting them in i.e. discord with 0 context is very confusing, I was wondering if there was a message that got deleted from the source slack at first

**Comment by wvffle** [comment]
Telegram also supports replies

**Comment by nicolas17** [comment]
Discord now has replies.

**Comment by patcon** [comment]
thanks @nicolas17! Updated issue body

**Comment by alex88** [comment]
Are you planning on supporting MS teams as well?

**Comment by patcon** [comment]
Heh not really sure I'd call these "plans", since I'm not working on anything here atm, but I added your suggestion to the checklist for tracking :)

**Comment by ys-chung** [comment]
Discord announced threads [will launch tomorrow](https://twitter.com/discord/status/1419721721519689733).

**Comment by qstambaugh** [comment]
At the very least, on the XMPP side, would we be able to get the message that's being replied to in a quote above the message? Example:

`> Message being replied to as a quote instead`
`The new message that's a reply to the quoted message above`

If someone is replying to a message on Discord or Matrix, it just shows the new message on XMPP with no indication that it's a reply or anything.

**Comment by glasgowm148** [comment]
Bumping this as Discord replies are lacking and probably the most needed feature on that platform as it makes the Discord side ugly and hard to read. I'm aware webhooks do not support replies, ([discussion here](https://github.com/discord/discord-api-docs/discussions/3282)) but there seems to be another implementation that handles them a bit better. 

The [Connections Bot](https://combot.xyz/) uses buttons instead and looks cleaner:

![image](https://user-images.githubusercontent.com/19927264/202471482-5f34f285-1cca-4492-be24-6ad7b34d6b4e.png)

**Comment by unode** [comment]
Matrix to Mattermost threads are working but Mattermost to Matrix messages are added off-thread.

Does anyone know if additional configuration is required to have symmetric support for this feature or if it's simply not implemented yet?

**Comment by patcon** [comment]
Thanks @locness3 @ys-chung -- updated the original message about discord and matrix threads!

@unode Sorry, not sure about bidirectional threading, but each direction is implemented separately -- incoming and outgoing need to be working. Someone may have gotten one direction working, without the other being done. You'll need to look into it, or maybe create a new issue if you think it's a regression (that it worked before).

**Comment by patcon** [comment]
Added mention and links to MSTeam code stub too

**Comment by saerdnaer** [comment]
@patcon Could you update rocket chat and matrix in the list above, they both support threads now since a few years...

**Comment by patcon** [comment]
Thanks @saerdnaer 

If someone wants to copy the above into a wiki page (and bonus, update it), then I'd be happy to link out from the issue body, so it's no longer outdated info
